### ISSUE TITLE
Any plan to use $jsonSchema?

### ISSUE THREAD

**Issue by shahidcodes** [Automattic/mongoose#11162]
Is there any plan use $jsonSchema for schema validation since MongoDB now supports it out of the box?

**Comment by vkarpov15** [comment]
We're not likely to make this change any time in the next few years. Since `$jsonSchema` is stored on the MongoDB server but Mongoose schemas are stored on the application, switching Mongoose to using `$jsonSchema` would be a massive change that would likely break a lot of codebases.

**Comment by veeramarni** [comment]
@vkarpov15 Can we use mix of both? Lets say i have a field with type `Mixed` and want to use JSON Schema fro that specific filed only.

**Comment by vkarpov15** [comment]
It is possible, we'll reopen this and consider it for a future release.

**Comment by Uzlopak** [comment]
Atleast it should be possible to pass the jsonschema to the mongoose schema or create the jsonschema from the mongoose schema .

**Comment by veeramarni** [comment]
Any example how to pass `jsonschema` to the mongoose schema?

**Comment by Uzlopak** [comment]
I checked a month ago. There is currently no opportunity to pass the jsonschema from the mongoose schema to createCollection

**Comment by GO-DIE** [comment]
非常期待这个功能得到支持

**Comment by webarthur** [comment]
A possible workaround:

```javascript
const mongoose = require('mongoose')
require('mongoose-schema-jsonschema')(mongoose)

const { Schema } = mongoose

const schema = new Schema({

  name: {
    type: String,
    required: [true, 'Please, fill your name.'],
    minlength: 2,
    maxlength: 120,
  },

  passwod: {
    type: String,
    required: [true, 'Please, fill your password.'],
    minlength: 8,
    maxlength: 40,
  },

  age: Number,

  active: {
    type: Boolean,
    // default: true, // issue here
  },

  hobbies: [String],

}, { timestamps: { createdAt: false, updatedAt: false } })

const jsonSchema = schema.jsonSchema()

schema.on('init', () => {
  const db = mongoose.connection

  db.once('open', () => {
    db.db.command({
      collMod: 'users',
      validator: {
        $jsonSchema: jsonSchema
      },
      validationLevel: 'moderate',
      validationAction: 'error'
    })
    .then(() => {
      console.log('Validation has been set up successfully')
    })
    .catch((error) => {
      console.error('Error configuring schema validation:', error)
    })
  })
})

const User = mongoose.model('user', schema)

module.exports = User
```

The json schema generated:

```javascript
{
  type: 'object',
  properties: {
    name: { type: 'string', minLength: 2, maxLength: 120 },
    passwod: { type: 'string', minLength: 8, maxLength: 40 },
    age: { type: 'number' },
    active: { type: 'boolean' },
    hobbies: { type: 'array', items: [Object] },
    _id: { type: 'string', pattern: '^[0-9a-fA-F]{24}$' }
  },
  required: [ 'name', 'passwod' ]
}
```

But there are some issues here. According to the [MongoDB documentation](https://www.mongodb.com/docs/manual/reference/operator/query/jsonSchema/#omissions), it is not possible to use some properties, such as default and format. How could this be resolved?

**Comment by harisvsulaiman** [comment]
@vkarpov15 can we revisit this

**Comment by vkarpov15** [comment]
@harisvsulaiman what are you suggesting? Do you want Mongoose to automatically apply JSON schema on MongoDB server based on Mongoose schema? Would a plugin be sufficient?

**Comment by harisvsulaiman** [comment]
> @harisvsulaiman what are you suggesting? Do you want Mongoose to automatically apply JSON schema on MongoDB server based on Mongoose schema? Would a plugin be sufficient?

@vkarpov15 yes a schema generated from mongoose schema which is applied on boot . 

I think a plugin is the best way to go about it.  It's easier for people to optin.

**Comment by harisvsulaiman** [comment]
@vkarpov15 we could create a plugin with above mentioned package and strip out the unsupported fields

**Comment by harisvsulaiman** [comment]
@vkarpov15 https://github.com/plexinc/papr seems to use the jsonschema spec

**Comment by LvChengbin** [comment]
Mongoose only validates data before sending it to MongoDB, but I need to verify the validity of data after updates as well. Therefore, support for $jsonSchema is essential.
Specifically, I need to use the `anyOf` keyword to ensure the update result is valid. And I don't think Mongoose needs to support `$jsonSchema` internally, it would be helpful simply to allow passing MongoDB's `[CreateCollectionOptions](https://mongodb.github.io/node-mongodb-native/4.9/interfaces/CreateCollectionOptions.html)` directly to `Mongoose.createCollection`.

**Comment by KTBsomen** [comment]
mongoose schema is very good so this is not a nessesary but nice to have

**Comment by vkarpov15** [comment]
Will be shipping `Schema.prototype.toJSONSchema` in 8.10, so you'll be able to do:

```javascript
await db.createCollection(collectionName, {
  validator: {
    $jsonSchema: schema.toJSONSchema({ useBsonType: true })
  }
});
```
